### PR TITLE
Include Signal-JEPA models

### PR DESCRIPTION
**Pull request by PierreGtch**
Add models from Signal-JEPA  (https://www.doi.org/10.3217/978-3-99161-014-4-003)

The `SignalJEPA_Contextual`, `SignalJEPA_PostLocal` and `SignalJEPA_PreLocal` architectures are meant for classification so they integrate well into the braindecode API.

However, the `SignalJEPA` arch was introduced for self-supervised pre-training and its output shape depends on the input shape. Not sure how/if we should integrate it.

### CODE REVIEW

**Comment by PierreGtch** [comment]
CC @tomMoral @MTangermann

File: braindecode/models/signal_jepa.py [hunk 1]
```diff
@@ -0,0 +1,994 @@
...
+        desc += f" | Overlap of {rf - ds_factor} samples"
+        if n_times is not None:
+            n_times_out = self.n_times_out(n_times)
+            desc += f" | {n_times_out} encoded samples/trial"
+
+        n_features = [
+            f"{dim}*{rf}" for dim, rf in zip([1] + list(dims), receptive_fields)
+        ]
+        desc += f" | #features/sample at each layer (n_channels*n_times): [{', '.join(n_features)}] = {[eval(x) for x in n_features]}"
+        return desc
```
Comment (bruAristimunha) [inline braindecode/models/signal_jepa.py:255]:
```suggestion
```

we already have torchinfo, no?

File: braindecode/models/signal_jepa.py [hunk 2]
```diff
@@ -0,0 +1,994 @@
...
+    The difference with a regular :class:`nn.Embedding` is that the embedding
+    vectors are initialized with a positional encodding of the channel locations.
+
+    Parameters
+    ----------
+        locs: list of (list of float or None)
+            List of the n-dimensions locations of the EEG channels.
+        embedding_dim: int
+            Dimensionality of the embedding vectors. Must be a multiple of the number
+            of dimensions of the channel locations.
```
Comment (bruAristimunha) [inline braindecode/models/signal_jepa.py:270]:
```suggestion
    locs: list of (list of float or None)
        List of the n-dimensions locations of the EEG channels.
    embedding_dim: int
        Dimensionality of the embedding vectors. Must be a multiple of the number
        of dimensions of the channel locations.
```

File: braindecode/models/signal_jepa.py [hunk 3]
```diff
@@ -0,0 +1,994 @@
...
+                        self.weight[
+                            i, j * self.coord_dim : (j + 1) * self.coord_dim
+                        ].copy_(
+                            pos_encode_contineous(
+                                x,
+                                0,
+                                10 * self.x_max,
+                                self.coord_dim,
+                                device=self.weight.device,
+                            ),
```
Comment (bruAristimunha) [inline braindecode/models/signal_jepa.py:310]:
```suggestion
    def __init__(self, channel_locations: list[list[float] | None], embedding_dim: int, **kwargs):
        # Determine the coordinate ranges for each dimension.
        # If a channel location has length 12, assume that indices 3:6 contain the coordinates.
        self.coordinate_ranges = [
            (min(coords), max(coords))
            for coords in zip(
                *[
                    loc[3:6] if len(loc) == 12 else loc
                    for loc in channel_locations
                    if loc is not None
                ]
            )
        ]
        # Compute the global min and max for the channel coordinates.
        channel_mins, channel_maxs = zip(*self.coordinate_ranges)
        global_min = min(channel_mins)
        global_max = max(channel_maxs)
        self.max_abs_coordinate = max(abs(global_min), abs(global_max))
        # Calculate how many embedding dimensions are allocated per coordinate.
        self.embedding_dim_per_coordinate = embedding_dim // len(self.coordinate_ranges)
        self.channel_locations = list(channel_locations)

        assert embedding_dim % len(self.coordinate_ranges) == 0
        super().__init__(len(channel_locations), embedding_dim, **kwargs)

    def reset_parameters(self):
        for i, channel_loc in enumerate(self.channel_locations):
            if channel_loc is None:
                nn.init.zeros_(self.weight[i])
            else:
                for coord_index, (coord_value, coord_range) in enumerate(zip(channel_loc, self.coordinate_ranges)):
                    coord_min, coord_max = coord_range  # Unpack the min and max of the coordinate range
                    with torch.no_grad():
                        self.weight[
                            i, coord_index * self.embedding_dim_per_coordinate : (coord_index + 1) * self.embedding_dim_per_coordinate
                        ].copy_(
                            pos_encode_contineous(
                                coord_value,
                                0,
                                10 * self.max_abs_coordinate,
                                self.embedding_dim_per_coordinate,
                                device=self.weight.device,
                            ),
                        )
```

File: braindecode/models/signal_jepa.py [hunk 4]
```diff
@@ -0,0 +1,994 @@
...
+            Number of dimensions to use to encode the temporal position of the patch.
+        ch_names: list[str]
+            List of all the EEG channel names that could be encountered in the data.
+        ch_locs: list of list of float or 2d array
+            List of the n-dimensions locations of the EEG channels.
+        sfreq_features: float
+            The "downsampled" sampling frequency returned by the feature encoder.
+        spat_kwargs: dict
+            Additional keyword arguments to pass to the :class:`nn.Embedding` layer used to
+            embed the channel names.
```
Comment (bruAristimunha) [inline braindecode/models/signal_jepa.py:332]:
```suggestion
    spat_dim: int
        Number of dimensions to use to encode the spatial position of the patch,
        i.e. the EEG channel.
    time_dim: int
        Number of dimensions to use to encode the temporal position of the patch.
    ch_names: list[str]
        List of all the EEG channel names that could be encountered in the data.
    ch_locs: list of list of float or 2d array
        List of the n-dimensions locations of the EEG channels.
    sfreq_features: float
        The "downsampled" sampling frequency returned by the feature encoder.
    spat_kwargs: dict
        Additional keyword arguments to pass to the :class:`nn.Embedding` layer used to
        embed the channel names.
```

File: braindecode/models/signal_jepa.py [hunk 5]
```diff
@@ -0,0 +1,994 @@
...
+    """
+
+    max_seconds: float = 600.0  # 10 minutes
```
Comment (bruAristimunha) [inline braindecode/models/signal_jepa.py:335]:
```suggestion
    max_seconds: float = 600.0  # 10 minutes
```

should be on the init this one, no?

File: braindecode/models/signal_jepa.py [hunk 6]
```diff
@@ -0,0 +1,994 @@
...
+        spat_kwargs = spat_kwargs or {}
+        ch_locs_plus_ukn = [None] + list(ch_locs)
+        self.ch_names = ch_names
+        self.pos_encoder_spat = ChannelEmbedding(
+            ch_locs_plus_ukn, spat_dim, **spat_kwargs
+        )  # (batch_size, n_channels, spat_dim)
+        self.spat_dim = spat_dim
+        self.time_dim = time_dim
+        self.max_n_times = int(self.max_seconds * sfreq_features)
+        self.encoding_time = torch.zeros(0, dtype=torch.float32, requires_grad=False)
```
Comment (bruAristimunha) [inline braindecode/models/signal_jepa.py:358]: just improve the variable names and comments here and

File: braindecode/models/signal_jepa.py [hunk 7]
```diff
@@ -0,0 +1,994 @@
...
+        )
+        # Channel pos. encoding
+        pos_encoding[:, :, :, : self.spat_dim] = self.pos_encoder_spat(ch_idxs)[
+            :, :, None, :
+        ]
+        # Temporal pos. encoding
+        self._check_encoding_time(n_times)
+        _ = pos_encoding[:, :, :, self.spat_dim : self.spat_dim + self.time_dim].copy_(
+            self.encoding_time[None, None, :n_times, :],
+        )
```
Comment (bruAristimunha) [inline braindecode/models/signal_jepa.py:407]:
```suggestion
        _ = pos_encoding[:, :, :, self.spat_dim : self.spat_dim + self.time_dim].copy_(
            self.encoding_time[None, None, :n_times, :],
        )
```

Why do we need copy here?
```suggestion
        _ = pos_encoding[:, :, :, self.spat_dim : self.spat_dim + self.time_dim].copy_(
            self.encoding_time[None, None, :n_times, :],
        )
```

File: braindecode/models/signal_jepa.py [hunk 8]
```diff
@@ -0,0 +1,994 @@
...
+            return (
+                torch.tensor(
+                    [self.ch_names.index(c) for c in self.fixed_ch_names]
+                ).unsqueeze(0)
+                + 1  # 0 is reserved for unknown channel
+            )
+        return batch["ch_idxs"]
+
+
+def n_times_out(conv_layers_spec, n_times):
```
Comment (bruAristimunha) [inline braindecode/models/signal_jepa.py:435]:
```suggestion
def _n_times_out(conv_layers_spec, n_times):
```

File: braindecode/models/signal_jepa.py [hunk 9]
```diff
@@ -0,0 +1,994 @@
...
+    # it would be equal to n_times//ds_factor without edge effects:
+    n_times_out_ = n_times
+    for _, width, stride in conv_layers_spec:
+        n_times_out_ = int((n_times_out_ - width) / stride) + 1
+    return n_times_out_
+
+
+def get_out_emb_dim(conv_layers_spec, n_times, n_spat_filters=4):
```
Comment (bruAristimunha) [inline braindecode/models/signal_jepa.py:443]:
```suggestion
def _get_out_emb_dim(conv_layers_spec, n_times, n_spat_filters=4):
```

File: braindecode/models/signal_jepa.py [hunk 10]
```diff
@@ -0,0 +1,994 @@
...
+    n_time_out = n_times_out(conv_layers_spec, n_times)
+    emb_dim = conv_layers_spec[-1][0]
+    return n_spat_filters * n_time_out * emb_dim
+
+
+def get_separable_clf_layer(
```
Comment (bruAristimunha) [inline braindecode/models/signal_jepa.py:449]:
```suggestion
def _get_separable_clf_layer(
```

File: braindecode/models/signal_jepa.py [hunk 11]
```diff
@@ -0,0 +1,989 @@
...
+        self.conv = nn.Conv2d(1, n_spat_filters, (n_chs, 1), **kwargs)
+        self.n_spat_filters = n_spat_filters
+
+    def forward(self, X):
+        assert X.ndim == 3
+        batch_size, n_chs, n_times = X.shape
+        X = X.view(batch_size, 1, n_chs, n_times)
+        X = self.conv(X)
+        X = X.view(batch_size, self.n_spat_filters, n_times)
+        return X
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/models/signal_jepa.py:100]: with einops

File: braindecode/models/modules.py [hunk 12]
```diff
@@ -912,3 +912,18 @@ def forward(self, x):
...
         return super(Conv2dWithConstraint, self).forward(x)
+
+
+class ToDevice(nn.Module):
+    def __init__(self, device=None):
+        super().__init__()
+        self.p = nn.Parameter(torch.empty((), device=device))
+
+    def forward(self, x: torch.Tensor):
+        return x.to(self.p.device)
```
[the remark this answers is not in the document]
Comment (PierreGtch) [reply braindecode/models/modules.py:923]: Not needed anymore because I removed torchtext. I will remove it
(But it's a nice solution I found to move newly created tensors to the same device as the rest of the model!)

File: braindecode/models/signal_jepa.py [hunk 13]
```diff
@@ -0,0 +1,989 @@
...
+
+
+class UnflattenTokens(nn.Module):
+    def __init__(self, n_chans: int):
+        super().__init__()
+        self.n_chans = n_chans
+
+    def forward(self, x: torch.Tensor):
+        batch_size, _, emb_dim = x.shape
+        return x.view(batch_size, 1, self.n_chans, -1, emb_dim)
```
[the remark this answers is not in the document]
Comment (PierreGtch) [reply braindecode/models/signal_jepa.py:83]: Cool! I didn't know we could do that

File: braindecode/models/signal_jepa.py [hunk 14]
```diff
@@ -0,0 +1,995 @@
...
+        def block(
+            n_in,
+            n_out,
+            k,
+            stride,
+            is_layer_norm=False,
+            is_group_norm=False,
+            conv_bias=False,
+        ):
+
```
[the remark this answers is not in the document]
Comment (PierreGtch) [reply braindecode/models/signal_jepa.py]: It avoids copying the parameters like activation, drop_prob dim... The code is more compact

File: braindecode/models/signal_jepa.py (hunk 7 above)
Comment (PierreGtch) [reply braindecode/models/signal_jepa.py:382]: I don't remember exactly why but I needed it

File: braindecode/models/signal_jepa.py (hunk 1 above)
Comment (PierreGtch) [reply braindecode/models/signal_jepa.py:230]: This gives more details (receptive field, downsampling factor, etc.)

**Review by bruAristimunha — APPROVED** [review]
just missing more tests and when you can merge!

**Comment by PierreGtch** [comment]
Awesome, thanks for the figures @bruAristimunha !!
I added the tests

**Comment by bruAristimunha** [comment]
Thank you @PierreGtch, amazing to have your model here ;)

**Review by bruAristimunha — APPROVED** [review]
